### PR TITLE
Revert PR #44

### LINKED ISSUE #45: Module fails if subnet uses default "main" routing table
## Describe the Bug
If one of the subnets in one of the VPCs is using the default "main" routing table, the module fails.

## Steps to Reproduce
1. Create a new VPC
2. Create new subnets in the VPC without associating route tables with them
3. Use the VPC as the "accepter"

```
Error: query returned no results. Please change your search criteria and try again

  on .terraform/modules/vpc_peering/accepter.tf line 64, in data "aws_route_table" "accepter":
  64: data "aws_route_table" "accepter" {
```

## Additional Context

This bug was introduced by #44 which incorporated #31 but still did not really fix #30.

I propose eliminating the option to filter subnets because it never really worked. If someone wants it back, they can submit a PR to implement it properly.

#### Sub-problem 1A: Looking up the route table for a subnet

You cannot find the route table for a subnet via 

```hcl
data "aws_route_table" "subnet" {
  subnet_id = local.subnet_id
}
```

because this will throw an error if the subnet does not have an associated route table and is instead using the default "main" table. That is the source of this bug. 

Because there is no way to make this conditional, we simply cannot use `data "aws_route_table"`.

#### Sub-problem 1B: Looking up the subnet for a route table

You could try looking up all the route tables and filtering them by associated subnets, but
- this is hard, because you have to treat the "main" route table specially and match it to the subnets that have no associated route table, and
- what do you do if you one route table is associated with both an included and an excluded subnet?

##### Conclusion: you might as well add routes to all the route tables, because you cannot promise better restrictions
After all, the entire VPC might have only one, default route table.

### Problem 2: Why are we looking up subnets by tag, anyway? 

The only meaningful restriction we can make is to limit routing to some portion of the VPC's CIDR, so we might as well pass explicit CIDRs for each VPC and just route those in all routing tables. It is not at all clear that selecting the CIDRs by subnet tag is more convenient than explicitly specifying them (or having the root module look them up any way it wants to).

### PR DESCRIPTION
**Pull request by Benbentwo**
## what
* Reverts PR #44 which introduced a bug when using default routing table

## references
* Fixes #45 
* Reopens https://github.com/cloudposse/terraform-aws-vpc-peering-multi-account/issues/30

### CODE REVIEW

File: accepter.tf [hunk 1]
```diff
@@ -64,36 +64,31 @@ locals {
...
   accepter_aws_route_table_ids_count     = length(local.accepter_aws_route_table_ids)
-  accepter_cidr_block_associations       = try(flatten(data.aws_vpc.accepter.*.cidr_block_associations), [])
+  accepter_cidr_block_associations       = flatten(data.aws_vpc.accepter.*.cidr_block_associations)
   accepter_cidr_block_associations_count = length(local.accepter_cidr_block_associations)
 }
 
 # Create routes from accepter to requester
 resource "aws_route" "accepter" {
-  count                     = local.accepter_enabled ? local.accepter_aws_route_table_ids_count * local.requester_cidr_block_associations_count : 0
+  count                     = module.this.enabled ? local.accepter_aws_route_table_ids_count * local.requester_cidr_block_associations_count : 0
```
Comment (aknysh) [inline accepter.tf:82]:
```suggestion
  count                     = local.enabled ? local.accepter_aws_route_table_ids_count * local.requester_cidr_block_associations_count : 0
```

why are you changing from `local.accepter_enabled` to the global enabled?

File: accepter.tf [hunk 2]
```diff
@@ -64,36 +64,31 @@ locals {
...
   vpc_peering_connection_id = join("", aws_vpc_peering_connection.requester.*.id)
   depends_on = [
-    data.aws_route_table.accepter,
+    data.aws_route_tables.accepter,
     aws_vpc_peering_connection_accepter.accepter,
-    aws_vpc_peering_connection.requester
+    aws_vpc_peering_connection.requester,
   ]
-
-  timeouts {
```
Comment (aknysh) [inline accepter.tf:93]: why removing the timeouts?

**Review by aknysh — COMMENTED** [review]
please see comments

File: accepter.tf (hunk 2 above)
Comment (Benbentwo) [reply accepter.tf:93]: accidental replacement

**Review by Nuru — CHANGES_REQUESTED** [review]
The `aws_subnet_ids` resource has been deprecated, and we should use `aws_subnets` instead. 

Although `aws_route_table` has not been deprecated, we should probably use `aws_route_tables` instead, as it has the cleaner, more modern interface.

We can fix the original bug by looking up the default route table and using it if no route table is found to be associated with the subnet.

**Comment by Benbentwo** [comment]
I believe I addressed your comments.  using `aws_subnet_ids` now. 

We we're already using `aws_route_tables` in the PR so nothing changed there. Tested with one using default route table, and one without.

File: accepter.tf [hunk 3]
```diff
@@ -48,15 +48,17 @@ data "aws_vpc" "accepter" {
...
-data "aws_subnet_ids" "accepter" {
+data "aws_subnets" "accepter" {
   count    = local.accepter_count
   provider = aws.accepter
-  vpc_id   = local.accepter_vpc_id
-  tags     = var.accepter_subnet_tags
+  filter {
+    name   = "vpc-id"
+    values = [local.accepter_vpc_id]
+  }
```
Comment (Nuru) [inline accepter.tf:57]: Include [tags](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/subnets#tags)

File: accepter.tf [hunk 4]
```diff
@@ -48,46 +48,48 @@ data "aws_vpc" "accepter" {
...
 }
 
 locals {
-  accepter_subnet_ids       = try(distinct(sort(flatten(data.aws_subnet_ids.accepter.*.ids))), [])
+  accepter_subnet_ids       = try(local.accepter_enabled ? data.aws_subnets.accepter[0].ids : [], [])
```
Comment (Nuru) [inline accepter.tf:61]: Apparently I was wrong about sorting data with respect to `data` sources, which I guess are now read during the plan phase. (See what goes into the `count` on line 84.) Please revert, as `distinct` and `sort` add stability to the plans.

File: accepter.tf [hunk 5]
```diff
@@ -48,46 +48,48 @@ data "aws_vpc" "accepter" {
...
 
 # Lookup accepter route tables
-data "aws_route_table" "accepter" {
-  count     = local.accepter_enabled ? local.accepter_subnet_ids_count : 0
-  provider  = aws.accepter
-  subnet_id = element(local.accepter_subnet_ids, count.index)
+data "aws_route_tables" "accepter" {
+  count    = local.count
+  provider = aws.accepter
+  vpc_id   = local.accepter_vpc_id
```
Comment (Nuru) [inline accepter.tf:72]: This is going to add routes to all the route tables in the VPC. I may have said that was a good idea a year ago, but it strikes me as bad/dangerous now. 

I propose that you add a second `aws_route_tables` resource, and use one to filter based on subnet ID and the other to find the default route table. Only add routes to the subnets' route tables, and only add to the main route table if there is a subnet that does not have a route table.

**Review by Nuru — DISMISSED** [review]
Backtracking on removing `sort` for data from data sources, since it seems to work. See other comments.

File: accepter.tf [hunk 6]
```diff
@@ -48,46 +48,70 @@ data "aws_vpc" "accepter" {
...
+data "aws_route_tables" "accepter" {
+  for_each = toset(local.accepter_subnet_ids)
+  provider = aws.accepter
+  vpc_id   = local.accepter_vpc_id
+  filter {
+    name   = "association.subnet-id"
+    values = [each.key]
+  }
+}
+# If we had more subnets than routetables, we should update the default.
```
Comment (aknysh) [inline accepter.tf:84]: add an empty line between the data sources

File: accepter.tf [hunk 7]
```diff
@@ -124,3 +148,8 @@ output "accepter_accept_status" {
   value       = join("", aws_vpc_peering_connection_accepter.accepter.*.accept_status)
   description = "Accepter VPC peering connection request status"
 }
+
+output "accepter_route_table_map" {
+  value       = local.accepter_aws_rt_map
+  description = "Accepter VPC subnet to route table map"
```
Comment (aknysh) [inline accepter.tf:154]:
```suggestion
  description = "Map of accepter VPC subnet IDs to route table IDs"
```

File: accepter.tf [hunk 8]
```diff
@@ -124,3 +148,8 @@ output "accepter_accept_status" {
   value       = join("", aws_vpc_peering_connection_accepter.accepter.*.accept_status)
   description = "Accepter VPC peering connection request status"
 }
+
+output "accepter_route_table_map" {
```
Comment (aknysh) [inline accepter.tf:152]:
```suggestion
output "accepter_subnet_route_table_map" {
```

File: accepter.tf [hunk 9]
```diff
@@ -48,46 +48,70 @@ data "aws_vpc" "accepter" {
...
+    content {
+      name   = "tag:${filter.key}"
+      values = [filter.value]
+    }
+  }
 }
 
 locals {
-  accepter_subnet_ids       = try(distinct(sort(flatten(data.aws_subnet_ids.accepter.*.ids))), [])
+  accepter_subnet_ids       = try(local.accepter_enabled ? data.aws_subnets.accepter[0].ids : [], [])
```
Comment (aknysh) [inline accepter.tf:68]: do we need `try` here?
it's useful in some cases, but mostly it hides errors that we don't want to hide

**Review by aknysh — COMMENTED** [review]
@Benbentwo a few nitpicks

File: accepter.tf (hunk 9 above)
Comment (Benbentwo) [reply accepter.tf]: I agree, was thinking of that when writing it. kept in, but there's no need, it's removed now

File: accepter.tf [hunk 10]
```diff
@@ -55,25 +55,48 @@ data "aws_subnets" "accepter" {
     name   = "vpc-id"
     values = [local.accepter_vpc_id]
   }
+  dynamic "filter" {
+    for_each = var.accepter_subnet_tags
+    content {
+      name   = "tag:${filter.key}"
+      values = [filter.value]
+    }
+  }
```
Comment (Nuru) [inline accepter.tf:64]:
```suggestion
  tags = var.accepter_subnet_tags
```
See https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/subnets#tags

File: accepter.tf [hunk 11]
```diff
@@ -126,3 +149,8 @@ output "accepter_accept_status" {
   value       = join("", aws_vpc_peering_connection_accepter.accepter.*.accept_status)
   description = "Accepter VPC peering connection request status"
 }
+
+output "accepter_subnet_route_table_map" {
+  value       = local.accepter_aws_rt_map
+  description = "Map of accepter VPC subnet IDs to route table IDs"
+}
```
Comment (Nuru) [inline accepter.tf:156]:
```suggestion
}

```

**Review by Nuru — CHANGES_REQUESTED** [review]

File: accepter.tf (hunk 10 above)
Comment (Benbentwo) [reply accepter.tf:64]: totally misread that documentation

**Review by Nuru — APPROVED** [review]

**Review by aknysh — APPROVED** [review]

**Pull request merged**